### PR TITLE
Treat "multi-lets" as syntactic sugar

### PR DESCRIPTION
**Pull request by EggBaconAndSpam**
Removes any mention of "multi-lets" from the semantics, while adding a
note to the grammar that `let... let... in...` is to be treated as
syntactic sugar for `let... in let... in...`. The resulting standard is
equivalent to the old one, only arguably simpler and easier to follow.

Addresses https://github.com/dhall-lang/dhall-lang/issues/615; see
https://github.com/dhall-lang/dhall-lang/commit/022105eefee605f6e5b862b9d7816862b9db73e2
for the commit that originally introduced multi-lets into the standard.

### CODE REVIEW

File: standard/dhall.abnf [hunk 1]
```diff
@@ -709,33 +711,33 @@ type-selector = "(" whsp expression whsp ")"
...
     ; "{ foo : Integer, bar : Bool }"
     / "{" whsp record-type-or-literal whsp "}"
-    
+
     ; "< Foo : Integer | Bar : Bool >"
     ; "< Foo : Integer | Bar = True | Baz : Bool >"
     ; "< Foo | Bar : Bool >"
     / "<" whsp union-type-or-literal whsp ">"
-    
+
```
Comment (f-f) [inline standard/dhall.abnf]: I think we should not remove all this whitespace, see https://github.com/dhall-lang/dhall-lang/pull/379

File: standard/README.md [hunk 2]
```diff
@@ -211,10 +211,6 @@ t          : Naked label which could be any type of expression.
                              value of `t₀`.  At least one alternative is named
                              `x₁` with a type of `T₁`.
 
-
-let xs… in b                : A `let` definition with at least one bindings
```
Comment (singpolyma) [inline standard/README.md]: should this not stay and just say "a `let` definition" ?

File: standard/binary.md [hunk 3]
```diff
@@ -652,7 +652,7 @@ interpolated expressions:
     encode("a${b₀}c${d}e…x${y₀}z") = [ 18, "a", b₁, "c", d₁, "e", …, "x", y₁, "z" ]
 
 In other words: the amount of encoded elements is always an odd number, with the
-odd elements being strings and the even ones being interpolated expressions.  
+odd elements being strings and the even ones being interpolated expressions.
```
Comment (singpolyma) [inline standard/binary.md]: this change is not related and should not be included

File: standard/dhall.abnf [hunk 4]
```diff
@@ -616,36 +616,38 @@ import = import-hashed [ whsp as whsp1 (Text / Location) ]
 expression =
     ; "\(x : a) -> b"
       lambda whsp "(" whsp nonreserved-label whsp ":" whsp1 expression whsp ")" whsp arrow whsp expression
-    
+
```
Comment (singpolyma) [inline standard/dhall.abnf]: lots of unrelated whitespace changes in this file

File: standard/dhall.abnf (hunk 1 above)
Comment (EggBaconAndSpam) [reply standard/dhall.abnf]: Ups... well-caught!

File: standard/README.md (hunk 2 above)
Comment (EggBaconAndSpam) [reply standard/README.md]: That one defines a notation for induction on lets with multiple binders. Since with this change we only consider lets with a single binder we don't need any notation for induction.

File: standard/dhall.abnf (hunk 4 above)
Comment (EggBaconAndSpam) [reply standard/dhall.abnf]: Yeah, I'll fix this!

File: standard/binary.md (hunk 3 above)
Comment (EggBaconAndSpam) [reply standard/binary.md]: Fixed!

File: standard/dhall.abnf (hunk 1 above)
Comment (f-f) [reply standard/dhall.abnf]: It's not the first time I leave this comment, so maybe it'd be a good idea to add [apg-js2](https://github.com/ldthomas/apg-js2) to CI in the same way as `instaparse` (i.e. just trying to generate a parser from the grammar)

Comment (f-f) [reply standard/dhall.abnf]: (I'm not asking we do it here, just thinking out loud)

Comment (EggBaconAndSpam) [reply standard/dhall.abnf]: Yes! I certainly didn't expect abnf to be whitespace sensitive (though it is in any case polite to not include unrelated whitespace changes).

**Comment by Profpatsch** [comment]
Why is `dhall.abnf` a full diff?

**Comment by EggBaconAndSpam** [comment]
> Why is `dhall.abnf` a full diff?

I've got no idea... Should only be a diff of 2 lines...

**Comment by ocharles** [comment]
> Why is dhall.abnf a full diff?

Did the line-endings change?

**Comment by EggBaconAndSpam** [comment]
Yes, that's it...

**Comment by Gabriel439** [comment]
@EggBaconAndSpam: Try using a command-line tool to fix the endings (i.e. `dos2unix` or `unix2dos`)

**Comment by Gabriel439** [comment]
Another thing you can try is:

```bash
$ git checkout master standard/dhall.abnf  # Revert `dhall.abnf` to its contents on `master`
```

... and then reapply the desired change and commit that.

**Comment by EggBaconAndSpam** [comment]
I have tried that without success. The problems is that git changes the line endings when I check in changes to the file. My copy has CRLF endings (via unix2dos)...

**Comment by Gabriel439** [comment]
@EggBaconAndSpam: Let's start with the following steps to debug what is going on.  
Throughout this debugging process, do not use `dos2unix` or `unix2dos` to fix up the file.  Just follow these steps verbatim:

* First, undo all the changes you made to `.gitattributes` to get back to a clean slate

  ```bash
  $ git checkout master .gitattributes
  ```

* Paste here the contents of your `~/.gitconfig`

  I want to see if you have any settings in your `git` config that may be interfering

* Check out the original file using:

  ```bash
  $ git checkout master standard/dhall.abnf
  ```

* Paste the output of the following command to verify that the file you just reset is clean from `git`'s point of view:

  ```bash
  $ git diff standard/dhall.abnf
  ```

  If the `diff` is non-empty then stop here and note that.

* Edit your file to apply the desired change

* Use `ghci` to verify that all lines in your edited `./standard/dhall.abnf` still end with `\r\n`

  ```
  Prelude> text <- readFile "./standard/dhall.abnf" 
  Prelude> let carriageReturn "" = False; carriageReturn l = last l == '\r'
  Prelude> filter (not . carriageReturn) (lines text)
  []
  ```

  If not, then stop here and note that.

* Use `git diff` again to see if the change is limited to two lines from `git`'s point of view:

  ```bash
  $ git diff standard/dhall.abnf
  ```

  If the change is not limited to only two lines, then stop here and note that.

  Note that the `git diff` might display `^M` at the end of modified lines.  That's harmless and you can ignore those by adding this setting to your `.gitconfig`:

  ```
  [core]
          whitespace = cr-at-eol
  ```

* If `git diff` only displays a two-line diff, then commit the change since this means that you succeeded

**Comment by EggBaconAndSpam** [comment]
Thanks for taking the time to talk me through this, Gabriel! It turns out that I had `autocrlf = input` in my .gitconfig, which caused the problem. Setting it to `false` fixed it!

I particularly like the trick of using ghci to check the line endings ;) Turns out it's enough to just print the `text` variable, this escapes special characters and you can see `\r\n` everywhere!

**Comment by singpolyma** [comment]
Easier than GHCI to check line endings:

echo "./path/to/file as Text" | dhall

;)

**Comment by EggBaconAndSpam** [comment]
> Easier than GHCI to check line endings: echo "./path/to/file as Text" | dhall ;)

On my system the carriage returns are are printed as actual carriage returns, so this doesn't work...
